README.md
# **flat_dialog**

Modified, more updated fork of [rflutter_alert](https://pub.dev/packages/rflutter_alert).
FlatDialog is a simple method of creating various popup dialogs in Flutter. 

 ![Error Dialog](https://i.imgur.com/CNsf6jC.png)
# Features
- Simple animations
- Production ready dialog styles
- Recyclable and reusable dialogs
- Flat, easy to understand and beautiful

# Let's begin!

 1.  **Add dependency to your app:**
 ```yaml
dependencies:
	flat_dialog: ^1.0.0
```

2. **Use some of our ready dialogs!**
```dart
Future<bool> throwErrorDialog(BuildContext context) {
    return FlatDialog(
      context: context,
      type: DialogType.error, // We have more of these!
      title: "ERROR",
      desc: "We can't reach our server :(, try again later.",
      buttons: [
        FlatDialogButton(
		  width: 150,
		  onPressed: () => Navigator.pop(context),
          child: const Text(
            "Oh no!",
            style: TextStyle(color: Colors.white, fontSize: 20),
          ),
        )
      ],
    ).show();
  }
```

3.  **Customize your new flat dialog (with `style`)!** 
```dart
AlertStyle(
      animationStyle: AnimationStyle.fromTop, // How the dialog animates
      closeButtonVisible: false, // Does the [x] in the corner appear?
      isTapDismiss: true, // If you click outside the dialog, does it close?
      descStyle: TextStyle(fontWeight: FontWeight.bold), // How does the text look? 
      animationDuration: const Duration(milliseconds: 400), // How long does the animation last?
      alertBorder: RoundedRectangleBorder( // Do you want a border around the dialog?
        borderRadius: BorderRadius.circular(0.0),
        side: BorderSide(
          color: Colors.grey,
        ),
      ),
      titleStyle: TextStyle(
        color: Colors.red, // What color is the [title] text?
      ),
      boxConstraints: const BoxConstraints.expand( // Do you want constraints?
        width: 300,
      ),
    );
```

4. **Check out the example project for even more features!**

# Want to contribute?
- Open a pull request
- Make a new issue on GitHub.

Giant thank you to [Ratel](https://ratel.com.tr/ "Ratel") and [İzzet Öztürk](https://tr.linkedin.com/in/paltimoz "İzzet Öztürk") for providing the base for this project <3.

CHANGELOG.md
## [1.0.0] - 2020-06-28

* Release version

## [1.1.0] - 2020-07-08

* Added isDescScrollable so that the dialog support long descriptions

## [1.1.1] - 2020-07-08

* Added a different way of limiting dialog size.




test/flat_dialog_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:flat_dialog/flat_dialog.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//     expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//   });
// }


lib/src/flat_dialog_start.dart
import 'package:flutter/material.dart';

import 'flat_dialog_animations.dart';
import 'flat_dialog_button.dart';
import 'flat_dialog_style.dart';

enum DialogType { error, success, info, warning, none }

class FlatDialog {
  final BuildContext context;
  final DialogType type;
  final AlertStyle style;
  final Image image;
  final String title;
  final String desc;
  final Widget content;
  final List<FlatDialogButton> buttons;
  final Function closeFunction;

  FlatDialog({
    @required this.context,
    this.type,
    this.style = const AlertStyle(),
    this.image,
    @required this.title,
    this.desc,
    this.content,
    this.buttons,
    this.closeFunction,
  });

  /// Displays defined alert window
  Future<bool> show() async {
    return showGeneralDialog(
      context: context,
      pageBuilder: (BuildContext buildContext, Animation<double> animation,
          Animation<double> secondaryAnimation) {
        return _buildDialog();
      },
      barrierDismissible: style.isTapDismiss,
      barrierLabel: MaterialLocalizations.of(context).modalBarrierDismissLabel,
      barrierColor: style.overlayColor,
      transitionDuration: style.animationDuration,
      transitionBuilder: (
        BuildContext context,
        Animation<double> animation,
        Animation<double> secondaryAnimation,
        Widget child,
      ) =>
          _showAnimation(animation, secondaryAnimation, child),
    );
  }

// Will be added in next version.
  // void dismiss() {
  //   Navigator.pop(context);
  // }

  // Alert dialog content widget
  Widget _buildDialog() {
    return Center(
      child: ConstrainedBox(
        constraints: style.boxConstraints ??
            const BoxConstraints.expand(
                width: double.infinity, height: double.infinity),
        child: Center(
          child: SingleChildScrollView(
            child: AlertDialog(
              backgroundColor: style.backgroundColor ??
                  Theme.of(context).dialogBackgroundColor,
              shape: style.alertBorder ?? _defaultShape(),
              title: Center(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    _getCloseButton(),
                    Padding(
                      padding: EdgeInsets.fromLTRB(
                          10, style.closeButtonVisible ? 0 : 20, 10, 0),
                      child: Column(
                        children: <Widget>[
                          _getImage(),
                          const SizedBox(
                            height: 20,
                          ),
                          Text(
                            title,
                            style: style.titleStyle,
                            textAlign: TextAlign.center,
                          ),
                          SizedBox(
                            height: desc == null ? 5 : 10,
                          ),
                          if (desc == null)
                            Container()
                          else
                            style.isDescScrollable
                                // ? Container(
                                //     height: style.descHeight ?? 250,
                                //     child: SingleChildScrollView(
                                //       child: Text(
                                //         desc,
                                //         style: style.descStyle,
                                //         textAlign: TextAlign.center,
                                //       ),
                                //     ),
                                //   )
                                ? Container(
                                    constraints: BoxConstraints(
                                      minHeight:
                                          MediaQuery.of(context).size.height *
                                              0.025,
                                      maxHeight:
                                          style.descHeight,
                                    ),
                                    child: SingleChildScrollView(
                                      child: Text(
                                        desc,
                                        style: style.descStyle,
                                        textAlign: TextAlign.center,
                                      ),
                                    ),
                                  )
                                : Text(
                                    desc,
                                    style: style.descStyle,
                                    textAlign: TextAlign.center,
                                  ),
                          if (content == null) Container() else content,
                        ],
                      ),
                    )
                  ],
                ),
              ),
              contentPadding: style.buttonAreaPadding,
              content: Row(
                mainAxisAlignment: MainAxisAlignment.center,
                children: _getButtons(),
              ),
            ),
          ),
        ),
      ),
    );
  }

  ShapeBorder _defaultShape() {
    return RoundedRectangleBorder(
      borderRadius: BorderRadius.circular(10.0),
      side: BorderSide(
        color: Colors.blueGrey,
      ),
    );
  }

  Widget _getCloseButton() {
    return style.closeButtonVisible
        ? Padding(
            padding: const EdgeInsets.fromLTRB(0, 10, 10, 0),
            child: Container(
              alignment: FractionalOffset.topRight,
              child: Container(
                width: 20,
                height: 20,
                decoration: const BoxDecoration(
                  image: DecorationImage(
                    image: AssetImage(
                      'assets/images/close.png',
                      package: 'flat_dialog',
                    ),
                  ),
                ),
                child: Material(
                  color: Colors.transparent,
                  child: InkWell(
                    onTap: () {
                      Navigator.of(context).pop();
                      closeFunction();
                    },
                  ),
                ),
              ),
            ),
          )
        : Container();
  }

  // Returns defined buttons. Default: Cancel Button
  List<Widget> _getButtons() {
    final List<Widget> expandedButtons = [];
    if (buttons != null) {
      buttons.forEach(
        (button) {
          final Widget buttonWidget = Padding(
            padding: const EdgeInsets.only(left: 2, right: 2),
            child: button,
          );
          if (button.width != null && buttons.length == 1) {
            expandedButtons.add(buttonWidget);
          } else {
            expandedButtons.add(Expanded(
              child: buttonWidget,
            ));
          }
        },
      );
    } else {
      expandedButtons.add(
        Expanded(
          child: FlatDialogButton(
            onPressed: () => Navigator.pop(context),
            child: Text(
              "Cancel",
              style: TextStyle(color: Colors.white, fontSize: 20),
            ),
          ),
        ),
      );
    }

    return expandedButtons;
  }

// Returns alert image for icon
  Widget _getImage() {
    Widget response = image ?? Container();
    switch (type) {
      case DialogType.success:
        response = Image.asset(
          'assets/images/icon_success.png',
          package: 'flat_dialog',
        );
        break;
      case DialogType.error:
        response = Image.asset(
          'assets/images/icon_error.png',
          package: 'flat_dialog',
        );
        break;
      case DialogType.info:
        response = Image.asset(
          'assets/images/icon_info.png',
          package: 'flat_dialog',
        );
        break;
      case DialogType.warning:
        response = Image.asset(
          'assets/images/icon_warning.png',
          package: 'flat_dialog',
        );
        break;
      case DialogType.none:
        response = Container();
        break;
    }
    return response;
  }

// Shows alert with selected animation
  AnimatedWidget _showAnimation(Animation<double> animation,
      Animation<double> secondaryAnimation, Widget child) {
    if (style.animationStyle == AnimationStyle.fromRight) {
      return AnimationTransition.fromRight(
          animation, secondaryAnimation, child);
    } else if (style.animationStyle == AnimationStyle.fromLeft) {
      return AnimationTransition.fromLeft(animation, secondaryAnimation, child);
    } else if (style.animationStyle == AnimationStyle.fromBottom) {
      return AnimationTransition.fromBottom(
          animation, secondaryAnimation, child);
    } else if (style.animationStyle == AnimationStyle.grow) {
      return AnimationTransition.grow(animation, secondaryAnimation, child);
    } else if (style.animationStyle == AnimationStyle.shrink) {
      return AnimationTransition.shrink(animation, secondaryAnimation, child);
    } else {
      return AnimationTransition.fromTop(animation, secondaryAnimation, child);
    }
  }
}


lib/src/flat_dialog_style.dart
import 'package:flutter/material.dart';

enum AnimationStyle { fromRight, fromLeft, fromTop, fromBottom, grow, shrink }

class AlertStyle {
  final AnimationStyle animationStyle;
  final Duration animationDuration;
  final ShapeBorder alertBorder;
  final bool closeButtonVisible;
  final bool isTapDismiss;
  final bool isDescScrollable;
  final Color backgroundColor;
  final Color overlayColor;
  final TextStyle titleStyle;
  final TextStyle descStyle;
  final EdgeInsets buttonAreaPadding;
  final BoxConstraints boxConstraints;
  final double descHeight;

  const AlertStyle({
    this.animationStyle = AnimationStyle.fromLeft,
    this.alertBorder,
    this.isDescScrollable = false,
    this.descHeight,
    this.animationDuration = const Duration(milliseconds: 150),
    this.backgroundColor = const Color.fromRGBO(255, 255, 255, 50),
    this.overlayColor = Colors.black26,
    this.titleStyle = const TextStyle(
      color: Colors.black,
      fontWeight: FontWeight.w500,
      fontSize: 22.0,
    ),
    this.descStyle = const TextStyle(
      color: Colors.black,
      fontWeight: FontWeight.w500,
      fontSize: 18.0,
    ),
    this.buttonAreaPadding = const EdgeInsets.all(20.0),
    this.closeButtonVisible = true,
    this.isTapDismiss = true,
    this.boxConstraints,
  });
}


lib/src/flat_dialog_button.dart
/*
 * Credit to Ratel for parts of this code.
 */
import 'package:flutter/material.dart';

class FlatDialogButton extends StatelessWidget {
  final Widget child;
  final double width;
  final double height;
  final Color color;
  final Gradient gradient;
  final BorderRadius radius;
  final Function onPressed;

  /// DialogButton constructor
  const FlatDialogButton({
    Key key,
    this.child,
    this.width,
    this.height = 60,
    this.color,
    this.gradient,
    this.radius,
    this.onPressed,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      width: width,
      height: height,
      decoration: BoxDecoration(
        color: color ?? Colors.blueAccent.withOpacity(0.8),
        gradient: gradient,
        borderRadius: radius ?? BorderRadius.circular(6),
      ),
      child: Material(
        color: Colors.transparent,
        child: InkWell(
          onTap: () => {onPressed()},
          child: Center(
            child: child,
          ),
        ),
      ),
    );
  }
}

lib/src/flat_dialog_animations.dart
/*
 * Created by Ratel
 * https://ratel.com.tr
 * 
 * Modified by Bruno, most of the code is written by him, modified by Bruno.
 */
import 'package:flutter/material.dart';



class AnimationTransition {

  static SlideTransition fromRight(Animation<double> animation,
      Animation<double> secondaryAnimation, Widget child) {
    return SlideTransition(
      position: Tween<Offset>(
        begin: const Offset(1.0, 0.0),
        end: Offset.zero,
      ).animate(animation),
      child: child,
    );
  }

  static SlideTransition fromLeft(Animation<double> animation,
      Animation<double> secondaryAnimation, Widget child) {
    return SlideTransition(
      position: Tween<Offset>(
        begin: const Offset(-1.0, 0.0),
        end: Offset.zero,
      ).animate(animation),
      child: child,
    );
  }

  static SlideTransition fromTop(Animation<double> animation,
      Animation<double> secondaryAnimation, Widget child) {
    return SlideTransition(
      position: Tween<Offset>(
        begin: const Offset(0.0, -1.0),
        end: Offset.zero,
      ).animate(animation),
      child: child,
    );
  }

  static SlideTransition fromBottom(Animation<double> animation,
      Animation<double> secondaryAnimation, Widget child) {
    return SlideTransition(
      position: Tween<Offset>(
        begin: const Offset(0.0, 1.0),
        end: Offset.zero,
      ).animate(animation),
      child: child,
    );
  }

  static ScaleTransition grow(Animation<double> animation,
      Animation<double> secondaryAnimation, Widget child) {
    return ScaleTransition(
      scale: Tween<double>(
        begin: 0.0,
        end: 1.0,
      ).animate(
        CurvedAnimation(
          parent: animation,
          curve: const Interval(
            0.00,
            0.50,
            curve: Curves.fastOutSlowIn,
          ),
        ),
      ),
      child: child,
    );
  }

  static ScaleTransition shrink(Animation<double> animation,
      Animation<double> secondaryAnimation, Widget child) {
    return ScaleTransition(
      scale: Tween<double>(
        begin: 1.2,
        end: 1.0,
      ).animate(
        CurvedAnimation(
          parent: animation,
          curve: const Interval(
            0.50,
            1.00,
            curve: Curves.fastOutSlowIn,
          ),
        ),
      ),
      child: child,
    );
  }
}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flat_dialog/flat_dialog.dart';

void main() => runApp(RatelApp());

class RatelApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        appBar: AppBar(
          title: Text('FlatDialog'),
        ),
        body: PopupDialog(),
      ),
    );
  }
}

class PopupDialog extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Container(
            padding: const EdgeInsets.all(5),
            alignment: Alignment.center,
            width: 300,
            decoration:
                BoxDecoration(color: Colors.indigoAccent.withOpacity(0.8)),
            child: FlatButton(
              child: Text('Normal Dialog'),
              onPressed: () => _onBasicAlertPressed(context),
            ),
          ),
          SizedBox(
            height: 10,
          ),
          Container(
            padding: const EdgeInsets.all(5),
            alignment: Alignment.center,
            width: 300,
            decoration:
                BoxDecoration(color: Colors.indigoAccent.withOpacity(0.8)),
            child: FlatButton(
              child: Text('Error Dialog'),
              onPressed: () => _onAlertButtonPressed(context),
            ),
          ),
          SizedBox(
            height: 10,
          ),
          Container(
            padding: const EdgeInsets.all(5),
            alignment: Alignment.center,
            width: 300,
            decoration:
                BoxDecoration(color: Colors.indigoAccent.withOpacity(0.8)),
            child: FlatButton(
              child: Text('Dialog with Gradient Buttons'),
              onPressed: () => _onAlertButtonsPressed(context),
            ),
          ),
          const SizedBox(
            height: 10,
          ),
          Container(
            padding: const EdgeInsets.all(5),
            alignment: Alignment.center,
            width: 300,
            decoration:
                BoxDecoration(color: Colors.indigoAccent.withOpacity(0.8)),
            child: FlatButton(
              child: Text('Dialog with a Custom Style'),
              onPressed: () => _onAlertWithStylePressed(context),
            ),
          ),
          const SizedBox(
            height: 10,
          ),
          Container(
            padding: const EdgeInsets.all(5),
            alignment: Alignment.center,
            width: 300,
            decoration:
                BoxDecoration(color: Colors.indigoAccent.withOpacity(0.8)),
            child: FlatButton(
              child: Text('Dialog with an Custom Image'),
              onPressed: () => _onAlertWithCustomImagePressed(context),
            ),
          ),
          const SizedBox(
            height: 10,
          ),
          Container(
            padding: const EdgeInsets.all(5),
            alignment: Alignment.center,
            width: 300,
            decoration:
                BoxDecoration(color: Colors.indigoAccent.withOpacity(0.8)),
            child: FlatButton(
              child: Text('Dialog with some Custom Content'),
              onPressed: () => _onAlertWithCustomContentPressed(context),
            ),
          ),
        ],
      ),
    );
  }

  Future<bool> _onBasicAlertPressed(BuildContext context) {
    return FlatDialog(
            context: context,
            title: "DIALOG",
            desc: "Here is some example text.")
        .show();
  }

  Future<bool> _onAlertButtonPressed(BuildContext context) {
    return FlatDialog(
      context: context,
      type: DialogType.error,
      title: "ERROR",
      desc: "Some text for you to understand the layout :)",
      buttons: [
        FlatDialogButton(
          child: const Text(
            "Amazing!",
            style: TextStyle(color: Colors.white, fontSize: 20),
          ),
          onPressed: () => Navigator.pop(context),
          width: 150,
        )
      ],
    ).show();
  }

  Future<bool> _onAlertButtonsPressed(BuildContext context) {
    return FlatDialog(
      context: context,
      type: DialogType.warning,
      title: "DIALOG",
      desc: "Here is another dialog example.",
      buttons: [
        FlatDialogButton(
          child: Text(
            "FLAT",
            style: TextStyle(color: Colors.white, fontSize: 20),
          ),
          onPressed: () => Navigator.pop(context),
          color: Color.fromRGBO(0, 179, 134, 1.0),
        ),
        FlatDialogButton(
          child: const Text(
            "GRADIENT",
            style: TextStyle(color: Colors.white, fontSize: 20),
          ),
          onPressed: () => Navigator.pop(context),
          gradient: LinearGradient(colors: [
            Colors.redAccent,
            Colors.redAccent.withOpacity(0.6),
          ]),
        )
      ],
    ).show();
  }

  Future<bool> _onAlertWithStylePressed(BuildContext context) {
    final AlertStyle alertStyle = AlertStyle(
      animationStyle: AnimationStyle.fromTop,
      closeButtonVisible: false,
      isTapDismiss: false,
      isDescScrollable: true,
      descStyle: TextStyle(fontWeight: FontWeight.bold),
      animationDuration: const Duration(milliseconds: 400),
      alertBorder: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(0.0),
        side: BorderSide(
          color: Colors.grey,
        ),
      ),
      titleStyle: TextStyle(
        color: Colors.red,
      ),
      descHeight: 50
    );

    return FlatDialog(
      context: context,
      style: alertStyle,
      type: DialogType.info,
      title: "RFLUTTER ALERT",
      desc: """
Flutter is more awesome with RFlutter Alert.
""",
      buttons: [
        FlatDialogButton(
          child: Text(
            "Okay",
            style: const TextStyle(color: Colors.white, fontSize: 20),
          ),
          onPressed: () => Navigator.pop(context),
          color: const Color.fromRGBO(0, 179, 134, 1.0),
          radius: BorderRadius.circular(0.0),
        ),
      ],
    ).show();
  }

  Future<bool> _onAlertWithCustomImagePressed(BuildContext context) {
    return FlatDialog(
      context: context,
      title: "Alert with a different image.",
      desc: "Custom image example",
      image: Image.asset("assets/images/success.png"),
    ).show();
  }

  Future<bool> _onAlertWithCustomContentPressed(BuildContext context) {
    return FlatDialog(
        context: context,
        title: "LOGIN",
        content: Column(
          children: <Widget>[
            TextField(
              decoration: InputDecoration(
                icon: Icon(Icons.account_circle),
                labelText: 'E-Mail',
              ),
            ),
            TextField(
              obscureText: true,
              decoration: InputDecoration(
                icon: Icon(Icons.lock),
                labelText: 'Password',
              ),
            ),
          ],
        ),
        buttons: [
          FlatDialogButton(
            onPressed: () => Navigator.pop(context),
            child: Text(
              "LOGIN",
              style: TextStyle(color: Colors.white, fontSize: 20),
            ),
          )
        ]).show();
  }
}
